macaroonbakery/bakery.py

```diff
@@ -0,0 +1,280 @@
+                '[' + ','.join(encoded_discharges) + ']')
+            return self.get(url, **kwargs)
+
+        return response
+
+    def _discharge(self, macaroon):
+        '''Discharge a macaroon.
+
+        @param the macaroon object to discharge.
+        @returns the discharged macaroons as an array of macaroon object.
+        @raise DischargeError when an error occurs during the discharge
+            process.
+        '''
+        discharges = [macaroon]
+        first_party_location = macaroon.location
+        self._discharge_caveats(macaroon, discharges, first_party_location,
+                                macaroon)
+        return discharges
+
+    def _discharge_caveats(self, macaroon, discharges, first_party_location,
+                           primary_macaroon):
+        '''Gathers discharge macaroons for all the third party caveats
+           for the macaroon passed in.
+
+        @param macaroon the macaroon to discharge.
+        @param discharges the list of discharged macaroons.
+        @param first_party_location the location of the first party.
+        @param primary_macaroon used for signature of the discharge macaroon.
+        @raise DischargeError when an error occurs during the discharge
+            process.
+        '''
+        caveats = macaroon.third_party_caveats()
+        for caveat in caveats:
+            dm = self._get_discharge(first_party_location,
+                                     caveat[0],
+                                     caveat[1])
+            dm = primary_macaroon.prepare_for_request(dm)
+            discharges.append(dm)
+            self._discharge_caveats(dm, discharges, first_party_location,
+                                    primary_macaroon)
+
+    def _get_discharge(self, location, third_party_location, condition):
+        ''' Get the discharge macaroon from the third party location.
+
+        @param location the origin.
+        @param third_party_location where to get a discharge from.
+        @param condition associated  to the discharged macaroon.
+        @return a discharged macaroon.
+        @raise DischargeError when an error occurs during the discharge
+            process.
+        '''
+        headers = {'Bakery-Protocol-Version': 1}
+        if self._discharge_token is not None:
+            headers['Macaroons'] = self._discharge_token
+        payload = {'id': condition, 'location': location}
+
+        try:
+            response = requests.post(third_party_location + '/discharge',
+                                     headers=headers,
+                                     data=payload,
+                                     timeout=self._timeout)
+        except requests.RequestException as exc:
+            raise DischargeError(
+                'Unable to access discharge endpoint: {}'.format(exc.message))
+        if response.status_code == 200:
+            return self._extract_macaroon_from_response(response)
+        elif response.status_code == 401:
```

Comment:
I think it might be better to use similar logic to what's in the Go httpbakery package - ignore the actual status code, parse the response into an error and return that error inside a DischargeError if the code isn't "interaction required", thus passing unknown error values back intact.